public/app/features/trails/groop/parser.ts

```diff
@@ -0,0 +1,158 @@
+        // this group is too small
+        if (this.config.miscGroupKey) {
+          // put the values into the miscGroupValues
+          miscGroupValues.push(...group.values);
+        } else {
+          // put the values into the node itself
+          node.values.push(...group.values);
+        }
+        // keep track of the descendants
+        node.descendants += group.values.length;
+        // remove the group
+        node.groups.delete(key);
+      } else if (group.values.length > this.config.idealMaxGroupSize && level < this.config.maxDepth - 1) {
+        // this group is too big - see if we can split it into
+        // subgroups
+        group = this.parseStrings(group.values, level + 1);
+        node.groups.set(key, group);
+        node.descendants += group.descendants;
+      } else {
+        node.descendants += group.descendants;
+      }
+    }
+    if (this.config.miscGroupKey && miscGroupValues.length > 0) {
+      // looks like we have some values for a misc group
+      const group: Node = {
+        groups: new Map(),
+        values: miscGroupValues,
+        descendants: miscGroupValues.length,
+      };
+      node.groups.set(this.config.miscGroupKey, group);
+    }
+    return node;
+  }
+}
+
+// printTree returns a string representation of the tree.
+function printTree(node: Node): string {
+  return _printTree(node, 0);
+}
+
+function _printTree(node: Node, level = 0): string {
+  if (node === null) {
+    return '(null node)';
+  }
+  let out = '';
+  for (const [k, v] of node.groups.entries()) {
+    out += `g ${level} ${'\t'.repeat(level)}${k} (${v.descendants})\n`;
+    out += _printTree(v, level + 1);
+  }
+  for (const v of node.values) {
+    out += `v ${level} ${'\t'.repeat(level)}${v}\n`;
+  }
+  return out;
+}
+
+export interface Node {
+  groups: Map<string, Node>;
+  values: string[];
+  descendants: number;
+}
+
+// PrefixDelimited extracts the prefix of a string at the given level.
+function prefixDelimited(s: string, level: number): string {
+  let delimiterCount = 0;
+  for (let i = 0; i < s.length; i++) {
+    const char = s.charCodeAt(i);
+    // Check if the character is not a letter or digit (non-alphanumeric)
+    if (
+      !(
+        (
+          (char >= 0x30 && char <= 0x39) || // 0-9
+          (char >= 0x41 && char <= 0x5a) || // A-Z
+          (char >= 0x61 && char <= 0x7a) || // a-z
+          (char >= 0xc0 && char <= 0xd6) || // Latin-1 Supplement and Extended-A
+          (char >= 0xd8 && char <= 0xf6) || // Latin-1 Supplement and Extended-A
+          (char >= 0xf8 && char <= 0xff) || // Latin-1 Supplement and Extended-A
+          (char >= 0x0100 && char <= 0x017f)
+        ) // Latin Extended-A
+      )
+    ) {
+      delimiterCount++;
+      if (delimiterCount > level) {
+        return s.slice(0, i);
+      }
+    }
+  }
+  return s; // Return the entire string if the level is higher than the number of delimiters.
+}
```

Comment:
We will also need to consider the `:` character for alert rules metrics. Any of these metrics currently group into an an empty-string group.

Reply:
You are right. I'll add this to my todo list. Without that, I think this feature won't be `done`.

Reply:
Or maybe we can do it in a separate PR.

Reply:
Yeah. Currently the `-` does a good job of indicating "Other".

Given that the `:` character can appear anywhere (not just at the beginning), we can consider just skipping past  this character if it appears at the beginning, and using it as a prefix splitter afterward.